src/viam/sdk/tests/test_robot.cpp

```diff
@@ -171,30 +218,48 @@ BOOST_AUTO_TEST_CASE(test_get_operations) {
             auto ops = client->get_operations();
             auto mock_ops = mock_operations_response();
 
-            BOOST_TEST(vec_to_string_util(ops) == vec_to_string_util(mock_ops),
-                       boost::test_tools::per_element());
+            BOOST_TEST(ops == mock_ops, boost::test_tools::per_element());
         });
 }
 
+BOOST_AUTO_TEST_CASE(test_discovery) {
+    robot_client_to_mocks_pipeline([](std::shared_ptr<RobotClient> client,
+                                      MockRobotService& service) -> void {
+        auto components = mock_discovery_response();
+        auto component = components[0];
+        auto results = component.results->begin();
+        auto protos = mock_proto_discovery_response();
+        auto proto = protos[0];
+        auto proto_results = proto.results().fields().begin();
+
+        BOOST_CHECK_EQUAL(component.query.subtype, proto.query().subtype());
+        BOOST_CHECK_EQUAL(component.query.model, proto.query().model());
+        BOOST_CHECK_EQUAL(results->first, proto_results->first);
+        std::cout << component.results->begin()->second->proto_value().number_value() << " ------ "
+                  << proto.results().fields().begin()->second.string_value() << std::endl;
+        // BOOST_CHECK_EQUAL(results->second->proto_value().string_value(),
+        //                   proto_results->second.number_value());
```

Comment:
I've been chasing this down for a while to get to the bottom of a test failing. I think I finally narrowed it down enough to make a little sense of what is going on.

Goal: to get `results->second` and `proto_results->second` both to return "bar".
```
        auto components = mock_discovery_response();
        auto component = components[0];
        auto results = component.results->begin();
        auto protos = mock_proto_discovery_response();
        auto proto = protos[0];
        auto proto_results = proto.results().fields().begin();
        BOOST_CHECK_EQUAL(results->second->proto_value().string_value(),
                           proto_results->second.number_value());
```
For some reason, this does not work. (note: the `test_get_operations` test case is also failing, and I'm assuming it's because of this results value.) When printing out `results->second->proto_value().string_value()`, it is empty. After doing tests, I realized that if I set the value to the number 4, it works. `results->second->proto_value().number_value()` returns 4, and all tests pass. 

The issue is with this:
```
    AttributeMap results =
        std::make_shared<std::unordered_map<std::string, std::shared_ptr<ProtoType>>>();
    results->emplace("foo", std::make_shared<ProtoType>("bar"));
```
For some reason, making the pair like this does not work, but making the pair with `std::make_shared<ProtoType>(4)` works just fine.

I'm not sure why this is the case. The only difference I can find is that in `proto_type.cpp`, I see that number and string values are set differently:
```
        case 2: {
            const std::string s = boost::get<std::string>(proto_type_);
            *v.mutable_string_value() = s;
            break;
        }
        case 3: {
            const int i = boost::get<int>(proto_type_);
            v.set_number_value(i);
            break;
        }
```

Reply:
I have a solution for this. Short version is: we were creating a `ProtoType` not with a `std::string` but with a `const char*`, which we don't have a constructor for. The compiler was then interpreting this as a `bool`. This makes some sense (`const char*` is inherently truth-y or false-y) but also confuses me a bit because the `ProtoType` constructors are all explicit. In either case, adding a `const char*` constructor or explicitly creating the ProtoType with a `std::string({value})` solves the problem. Probably we should do both.